release-23.1: sql: rationalize the handling of comments and improve performance

rafiss opened the pull request:
Backport 1/1 commits from #104763.
Second commit cherry-picked from https://github.com/cockroachdb/cockroach/pull/105229

/cc @cockroachdb/release

Release justification: high value performance improvement

---

rebase of https://github.com/cockroachdb/cockroach/pull/95431
fixes https://github.com/cockroachdb/cockroach/issues/102851

Prior to this patch, we had a weird translation going in
`kv_catalog_comments`, which was causing the tree of objects to be
traversed 6 times (one per comment type), and then the rows to be
mapped one by one again in `pg_description` / `pg_shdescription`.

This made no sense - it should be possible to push a predicate on
`pg_description` down to a virtual index on `kv_catalog_comments`, and
this requires the column types to align and no computation to happen
in-between.

This commit simplifies all this by having `kv_catalog_comments`
responsible for preparing the data in the format expected by the `pg_`
tables.

The number of roundtrips needed for an ORM query that uses
pg_description dropped to 4, from 134.

```
goos: linux
goarch: amd64
cpu: Intel(R) Xeon(R) CPU @ 2.80GHz
                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │                sec/op                 │                   sec/op                    vs base                │
ORMQueries/introspection_description_join-24                           10687.1m ± 1%                                  144.1m ± 8%  -98.65% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │              roundtrips               │                 roundtrips                  vs base                │
ORMQueries/introspection_description_join-24                            134.000 ± 0%                                   4.000 ± 0%  -97.01% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │                 B/op                  │                    B/op                     vs base                │
ORMQueries/introspection_description_join-24                          3846.69Mi ± 0%                                 38.98Mi ± 6%  -98.99% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │               allocs/op               │                 allocs/op                   vs base                │
ORMQueries/introspection_description_join-24                           45566.9k ± 0%                                  433.3k ± 6%  -99.05% (p=0.000 n=10)
```

Release note: None

rafiss committed:
sql: rationalize the handling of comments

Prior to this patch, we had a weird translation going in
`kv_catalog_comments`, which was causing the tree of objects to be
traversed 6 times (one per comment type), and then the rows to be
mapped one by one again in `pg_description` / `pg_shdescription`.

This made no sense - it should be possible to push a predicate on
`pg_description` down to a virtual index on `kv_catalog_comments`, and
this requires the column types to align and no computation to happen
in-between.

This commit simplifies all this by having `kv_catalog_comments`
responsible for preparing the data in the format expected by the `pg_`
tables. It also adds kv_builtin_function_comments, so that the large
number of predefined comments on builtin functions don't cause joins on
the table to be very large in the common case.

The number of roundtrips needed for an ORM query that uses
pg_description dropped to 4, from 134.

```
goos: linux
goarch: amd64
cpu: Intel(R) Xeon(R) CPU @ 2.80GHz
                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │                sec/op                 │                   sec/op                    vs base                │
ORMQueries/introspection_description_join-24                           10687.1m ± 1%                                  144.1m ± 8%  -98.65% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │              roundtrips               │                 roundtrips                  vs base                │
ORMQueries/introspection_description_join-24                            134.000 ± 0%                                   4.000 ± 0%  -97.01% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │                 B/op                  │                    B/op                     vs base                │
ORMQueries/introspection_description_join-24                          3846.69Mi ± 0%                                 38.98Mi ± 6%  -98.99% (p=0.000 n=10)

                                             │ /tmp/tmp.0yUMg6OauX/bench.6abe1851d83 │ /tmp/tmp.0yUMg6OauX/bench.89a494eb3010b945ae5a2be7a057e50802b231c0 │
                                             │               allocs/op               │                 allocs/op                   vs base                │
ORMQueries/introspection_description_join-24                           45566.9k ± 0%                                  433.3k ± 6%  -99.05% (p=0.000 n=10)
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105218)
<!-- Reviewable:end -->

rafiss committed:
comprules: avoid hardcoding OIDs

Rather than hardcoding OIDs, now the query is modified so that the OID
is computed on the server-side. This improves cross-version
compatibility.

This also modifies the query to use the externally-usable pg_catalog,
rather than crdb_internal.

Release note: None

postamar commented:
Should the latest commit be front-ported?
